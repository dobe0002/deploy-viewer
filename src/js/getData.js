import axios from 'axios';
import each from 'lodash/each';
import sortBy from 'lodash/sortBy';
import map from 'lodash/map';
import isEmpty from 'lodash/isEmpty';
import { format } from 'date-fns';

class GetSpreadsheetData {
  constructor(docId, sheetNum = 1) {
    this.docId = docId;
    this.sheetNum = sheetNum;
    this.deployData = {};
    this.appGroups = {};

    this.url = `https://spreadsheets.google.com/feeds/list/${this.docId}/${this.sheetNum}/public/full?alt=json`;
  }

  getDeployData(done) {
    axios
      .get(this.url)
      .then(response => {
        const formatError = GetSpreadsheetData.verifyFormat(response.data);
        if (formatError !== null) {
          return done(formatError);
        }
        const deployData = this.formatData(response.data); // this sets this.deployData
        this.formatAppGroups(deployData); // this sets this.appGroups
        return done(null, deployData);
      })
      .catch(error => {
        // eslint-disable-next-line no-console
        console.error('ERROR in getDeployData:', error);
        return done(
          'There was an issue retrieving information from the Google Sheet.'
        );
      });
  }

  getGroups(done) {
    if (!isEmpty(this.appGroups)) {
      return done(null, this.appGroups);
    }

    return this.getDeployData((errorGetDeployData, deployData) => {
      const appGroups = this.formatAppGroups(deployData);
      this.appGroups = appGroups;
      return done(errorGetDeployData, appGroups);
    });
  }

  formatAppGroups(deployData) {
    const appGroups = {};
    each(deployData, app => {
      const { appGroup } = app;
      const { appGroupTitle } = app;
      if (!appGroups[appGroup]) {
        appGroups[appGroup] = appGroupTitle;
      }
    });
    this.appGroups = appGroups;
    return appGroups;
  }

  static verifyFormat(rawGoogleJson) {
    if (
      rawGoogleJson.feed === undefined ||
      rawGoogleJson.feed.entry === undefined ||
      rawGoogleJson.feed.entry.length === 0
    ) {
      return 'No data was sent from the Google spreadsheet.';
    }
    const testEntry = rawGoogleJson.feed.entry[0];

    if (testEntry.gsx$app === undefined) {
      return 'Application title is not available.';
    }
    if (testEntry.gsx$appcode === undefined) {
      return 'Application code is not available.';
    }
    if (testEntry.gsx$groupcode === undefined) {
      return 'Application group code is not available.';
    }
    if (testEntry.gsx$appgroup === undefined) {
      return 'Application group is not available.';
    }
    if (testEntry.gsx$date === undefined) {
      return 'Event date is not available.';
    }
    if (testEntry.gsx$eventcode === undefined) {
      return 'Event code is not available.';
    }
    if (testEntry.gsx$event === undefined) {
      return 'Event description is not available.';
    }
    return null;
  }

  formatData(rawGoogleJson) {
    let deployData = {};
    each(rawGoogleJson.feed.entry, (entry, i) => {
      // note Google skips the first row - it assumes it is a header row
      const appCode = entry.gsx$appcode.$t;
      if (deployData[appCode] === undefined) {
        // Add new app into the deployData object:
        deployData[appCode] = {
          title: entry.gsx$app.$t,
          appCode,
          appGroup: entry.gsx$groupcode.$t,
          appGroupTitle: entry.gsx$appgroup.$t,
          status: []
        };
      }

      // Add status
      const date = entry.gsx$date.$t.match(/\d+/g);
      const year = parseInt(date[2], 10);
      const month = parseInt(date[0], 10) - 1;
      const day = parseInt(date[1], 10);
      const jsDate = new Date(year, month, day);

      deployData[appCode].status.push({
        row: i + 2,
        status: entry.gsx$eventcode.$t,
        statusDescription: entry.gsx$event.$t,
        dateString: format(jsDate, 'MM/dd/yyyy'),
        dateSort: parseInt(format(jsDate, 'yyyyMMdd'), 10),
        date: jsDate,
        note: entry.gsx$notes.$t
      });
    });
    // Sort each status
    each(deployData, app => {
      deployData[app.appCode].status = sortBy(
        app.status,
        event => event.dateString
      );
    });

    // convert back to array for easy sorting
    deployData = map(deployData, app => app);

    // Sort by application
    deployData = sortBy(deployData, app => app.title);

    this.deployData = deployData;
    return deployData;
  }
}

export default GetSpreadsheetData;
