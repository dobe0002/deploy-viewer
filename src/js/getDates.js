import each from 'lodash/each';
import { min, max, eachDayOfInterval, format } from 'date-fns';

const getStartEndDate = deployData => {
  const allDates = [];
  each(deployData, app => {
    each(app.status, status => {
      allDates.push(status.date);
    });
  });

  const minDate = min(allDates);
  const maxDate = max(allDates);
  return { min: minDate, max: maxDate };
};

const getDates = (minDate, maxDate) => {
  const JSDateArray = eachDayOfInterval({
    start: minDate,
    end: maxDate
  });
  return JSDateArray.map(date => {
    const newDate = new Date(date);
    newDate.setUTCHours(6);
    newDate.setUTCMinutes(0);
    newDate.setUTCSeconds(0);
    newDate.setUTCMinutes(0);
    return newDate;
  });
};
const convertStringToDate = dateString => {
  if (dateString === undefined) {
    return dateString;
  }
  const dateArray = dateString.split('-');
  const month = parseInt(dateArray[0], 10) - 1;
  const day = parseInt(dateArray[1], 10);
  const year = parseInt(dateArray[2], 10);
  return new Date(year, month, day);
};
const convertDateToString = date => {
  return format(date, 'MM-dd-yyyy');
};

export { getStartEndDate, getDates, convertStringToDate, convertDateToString };
