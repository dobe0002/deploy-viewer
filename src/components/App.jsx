import React from 'react';
import PropTypes from 'prop-types';
import {
  Route,
  HashRouter as Router,
  Switch,
  NavLink,
  Redirect
} from 'react-router-dom';

import Builds from './Builds';
import Timeline from './Timeline';
import Error from './Error';
import { convertStringToDate } from '../js/getDates';
import FourOhFour from './FourOhFour';

export default class App extends React.Component {
  buildNav() {
    const { googleId } = this.props;
    const googleSheetURL = `https://docs.google.com/spreadsheets/d/${googleId}/`;
    return (
      <nav className='navbar navbar-expand-lg navbar-light bg-light'>
        <span className='navbar-brand'>AppDev RAS/C Deploy Status</span>

        <div className='navbar' id='navbarNav'>
          <ul className='navbar-nav'>
            <li className='nav-item'>
              <NavLink
                exact
                activeClassName='active'
                className='nav-link'
                to='/deploys'
              >
                Deploys
              </NavLink>
            </li>
            <li className='nav-item'>
              <NavLink
                activeClassName='active'
                className='nav-link'
                to='/timeline'
              >
                Timeline
              </NavLink>
            </li>
            <li className='nav-item'>
              <a
                className='nav-link'
                href={googleSheetURL}
                target='_blank'
                rel='noreferrer noopener'
              >
                Google Sheet
              </a>
            </li>
          </ul>
        </div>
      </nav>
    );
  }

  render() {
    const basename = '';
    const { deployData, appGroups, error } = this.props;
    return (
      <Router basename={basename}>
        {this.buildNav()}
        <main>
          <Switch>
            {error !== '' && (
              <Route path='/' render={() => <Error error={error} />} />
            )}
            <Route
              path='/deploys//startDate/:startDate/endDate/:endDate'
              render={props => (
                <Builds
                  deployData={deployData}
                  appGroups={appGroups}
                  currentAppGroup={props.match.params.appGroup}
                  history={props.history}
                  startDate={convertStringToDate(props.match.params.startDate)}
                  endDate={convertStringToDate(props.match.params.endDate)}
                />
              )}
            />
            <Route
              path='/deploys/:appGroup/startDate/:startDate/endDate/:endDate'
              render={props => (
                <Builds
                  deployData={deployData}
                  appGroups={appGroups}
                  currentAppGroup={props.match.params.appGroup}
                  history={props.history}
                  startDate={convertStringToDate(props.match.params.startDate)}
                  endDate={convertStringToDate(props.match.params.endDate)}
                />
              )}
            />
            <Route
              path='/deploys/:appGroup?'
              render={props => (
                <Builds
                  deployData={deployData}
                  appGroups={appGroups}
                  currentAppGroup={props.match.params.appGroup}
                  history={props.history}
                />
              )}
            />
            <Route
              path='/timeline//startDate/:startDate/endDate/:endDate'
              render={props => (
                <Timeline
                  deployData={deployData}
                  appGroups={appGroups}
                  history={props.history}
                  startDate={convertStringToDate(props.match.params.startDate)}
                  endDate={convertStringToDate(props.match.params.endDate)}
                />
              )}
            />

            <Route
              path='/timeline/:appGroup/startDate/:startDate/endDate/:endDate'
              render={props => (
                <Timeline
                  deployData={deployData}
                  appGroups={appGroups}
                  history={props.history}
                  currentAppGroup={props.match.params.appGroup}
                  startDate={convertStringToDate(props.match.params.startDate)}
                  endDate={convertStringToDate(props.match.params.endDate)}
                />
              )}
            />
            <Route
              path='/timeline/:appGroup?'
              render={props => (
                <Timeline
                  deployData={deployData}
                  appGroups={appGroups}
                  history={props.history}
                  currentAppGroup={props.match.params.appGroup}
                />
              )}
            />
            <Route path='/' exact>
              <Redirect to='/deploys' />
            </Route>

            <Route path='/'>
              <FourOhFour />
            </Route>
          </Switch>
        </main>
      </Router>
    );
  }
}
App.propTypes = {
  deployData: PropTypes.arrayOf,
  appGroups: PropTypes.objectOf,
  googleId: PropTypes.string.isRequired,
  error: PropTypes.string
};
App.defaultProps = {
  deployData: [],
  appGroups: {},
  error: ''
};
