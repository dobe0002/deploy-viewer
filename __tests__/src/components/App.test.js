/* eslint-disable react/jsx-filename-extension */
import React from 'react';

import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { MemoryRouter } from 'react-router'; // https://medium.com/@antonybudianto/react-router-testing-with-jest-and-enzyme-17294fefd303
import App from '../../../src/components/App';
import deployData from '../../../__fixtures__/getDataFormat/sheetData';
import { convertStringToDate } from '../../../src/js/getDates';

configure({ adapter: new Adapter() });

const { axe, toHaveNoViolations } = require('jest-axe');

expect.extend(toHaveNoViolations);

describe('App  tests', () => {
  const sampleProps = {
    deployData,
    googleId: '1234567890'
  };
  test('Accessibility check', async () => {
    const div = document.createElement('div');
    div.id = sampleProps.id;
    document.body.appendChild(div);
    const wrapper = mount(
      <MemoryRouter initialEntries={['/deploys']}>
        <App
          deployData={sampleProps.deployData}
          googleId={sampleProps.googleId}
        />
      </MemoryRouter>,
      { attachTo: div }
    );
    const results = await axe(wrapper.html());
    expect(results).toHaveNoViolations();
  });

  test('Snapshot tests', () => {
    const wrapper = shallow(
      <App
        deployData={sampleProps.deployData}
        googleId={sampleProps.googleId}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  test('Deploy - no attributes', () => {
    const div = document.createElement('div');
    div.id = sampleProps.id;
    document.body.appendChild(div);
    const wrapper = mount(
      <MemoryRouter initialEntries={['/deploys']}>
        <App
          deployData={sampleProps.deployData}
          googleId={sampleProps.googleId}
        />
      </MemoryRouter>,
      { attachTo: div }
    );
    expect(wrapper.find('Builds')).toHaveLength(1);
    expect(wrapper.find('Timeline')).toHaveLength(0);
    expect(wrapper.find('Builds').prop('currentAppGroup')).toEqual('');
    expect(wrapper.find('Builds').prop('startDate')).toBeUndefined();
    expect(wrapper.find('Builds').prop('endDate')).toBeUndefined();
  });
  test('Deploy - app group', () => {
    const div = document.createElement('div');
    div.id = sampleProps.id;
    document.body.appendChild(div);
    const wrapper = mount(
      <MemoryRouter initialEntries={['/deploys/Foo']}>
        <App
          deployData={sampleProps.deployData}
          googleId={sampleProps.googleId}
        />
      </MemoryRouter>,
      { attachTo: div }
    );
    expect(wrapper.find('Builds')).toHaveLength(1);
    expect(wrapper.find('Timeline')).toHaveLength(0);
    expect(wrapper.find('Builds').prop('currentAppGroup')).toEqual('Foo');
    expect(wrapper.find('Builds').prop('startDate')).toBeUndefined();
    expect(wrapper.find('Builds').prop('endDate')).toBeUndefined();
  });
  test('Deploy - start/end dates', () => {
    const div = document.createElement('div');
    div.id = sampleProps.id;
    document.body.appendChild(div);
    const wrapper = mount(
      <MemoryRouter
        initialEntries={['/deploys//startDate/2019-01-01/endDate/2019-02-02']}
      >
        <App
          deployData={sampleProps.deployData}
          googleId={sampleProps.googleId}
        />
      </MemoryRouter>,
      { attachTo: div }
    );
    expect(wrapper.find('Builds')).toHaveLength(1);
    expect(wrapper.find('Timeline')).toHaveLength(0);
    expect(wrapper.find('Builds').prop('currentAppGroup')).toEqual('');
    expect(wrapper.find('Builds').prop('startDate')).toEqual(
      convertStringToDate('2019-01-01')
    );
    expect(wrapper.find('Builds').prop('endDate')).toEqual(
      convertStringToDate('2019-02-02')
    );
  });
  test('Deploy - start/end dates and app group', () => {
    const div = document.createElement('div');
    div.id = sampleProps.id;
    document.body.appendChild(div);
    const wrapper = mount(
      <MemoryRouter
        initialEntries={[
          '/deploys/Foo/startDate/2019-01-01/endDate/2019-02-02'
        ]}
      >
        <App
          deployData={sampleProps.deployData}
          googleId={sampleProps.googleId}
        />
      </MemoryRouter>,
      { attachTo: div }
    );
    expect(wrapper.find('Builds')).toHaveLength(1);
    expect(wrapper.find('Timeline')).toHaveLength(0);
    expect(wrapper.find('Builds').prop('currentAppGroup')).toEqual('Foo');
    expect(wrapper.find('Builds').prop('startDate')).toEqual(
      convertStringToDate('2019-01-01')
    );
    expect(wrapper.find('Builds').prop('endDate')).toEqual(
      convertStringToDate('2019-02-02')
    );
  });
  test('Timeline - no attributes', () => {
    const div = document.createElement('div');
    div.id = sampleProps.id;
    document.body.appendChild(div);
    const wrapper = mount(
      <MemoryRouter initialEntries={['/timeline']}>
        <App
          deployData={sampleProps.deployData}
          googleId={sampleProps.googleId}
        />
      </MemoryRouter>,
      { attachTo: div }
    );
    expect(wrapper.find('Builds')).toHaveLength(0);
    expect(wrapper.find('Timeline')).toHaveLength(1);
    expect(wrapper.find('Timeline').prop('currentAppGroup')).toEqual('');
    expect(wrapper.find('Timeline').prop('startDate')).toBeUndefined();
    expect(wrapper.find('Timeline').prop('endDate')).toBeUndefined();
  });
  test('Timeline - app group', () => {
    const div = document.createElement('div');
    div.id = sampleProps.id;
    document.body.appendChild(div);
    const wrapper = mount(
      <MemoryRouter initialEntries={['/timeline/Foo']}>
        <App
          deployData={sampleProps.deployData}
          googleId={sampleProps.googleId}
        />
      </MemoryRouter>,
      { attachTo: div }
    );
    expect(wrapper.find('Builds')).toHaveLength(0);
    expect(wrapper.find('Timeline')).toHaveLength(1);
    expect(wrapper.find('Timeline').prop('currentAppGroup')).toEqual('Foo');
    expect(wrapper.find('Timeline').prop('startDate')).toBeUndefined();
    expect(wrapper.find('Timeline').prop('endDate')).toBeUndefined();
  });
  test('Timeline - start/end dates', () => {
    const div = document.createElement('div');
    div.id = sampleProps.id;
    document.body.appendChild(div);
    const wrapper = mount(
      <MemoryRouter
        initialEntries={['/timeline//startDate/2019-01-01/endDate/2019-02-02']}
      >
        <App
          deployData={sampleProps.deployData}
          googleId={sampleProps.googleId}
        />
      </MemoryRouter>,
      { attachTo: div }
    );
    expect(wrapper.find('Builds')).toHaveLength(0);
    expect(wrapper.find('Timeline')).toHaveLength(1);
    expect(wrapper.find('Timeline').prop('currentAppGroup')).toEqual('');
    expect(wrapper.find('Timeline').prop('startDate')).toEqual(
      convertStringToDate('2019-01-01')
    );
    expect(wrapper.find('Timeline').prop('endDate')).toEqual(
      convertStringToDate('2019-02-02')
    );
  });
  test('Timeline - start/end dates and app group', () => {
    const div = document.createElement('div');
    div.id = sampleProps.id;
    document.body.appendChild(div);
    const wrapper = mount(
      <MemoryRouter
        initialEntries={[
          '/timeline/Foo/startDate/2019-01-01/endDate/2019-02-02'
        ]}
      >
        <App
          deployData={sampleProps.deployData}
          googleId={sampleProps.googleId}
        />
      </MemoryRouter>,
      { attachTo: div }
    );
    expect(wrapper.find('Builds')).toHaveLength(0);
    expect(wrapper.find('Timeline')).toHaveLength(1);
    expect(wrapper.find('Timeline').prop('currentAppGroup')).toEqual('Foo');
    expect(wrapper.find('Timeline').prop('startDate')).toEqual(
      convertStringToDate('2019-01-01')
    );
    expect(wrapper.find('Timeline').prop('endDate')).toEqual(
      convertStringToDate('2019-02-02')
    );
  });

  test('Show Error', () => {
    const div = document.createElement('div');
    div.id = sampleProps.id;
    document.body.appendChild(div);
    const wrapper = mount(
      <MemoryRouter initialEntries={['/timeline']}>
        <App
          deployData={sampleProps.deployData}
          googleId={sampleProps.googleId}
          error='An error has occurred!'
        />
      </MemoryRouter>,
      { attachTo: div }
    );
    expect(wrapper.find('Builds')).toHaveLength(0);
    expect(wrapper.find('Timeline')).toHaveLength(0);
    expect(wrapper.find('Error')).toHaveLength(1);
  });
});
