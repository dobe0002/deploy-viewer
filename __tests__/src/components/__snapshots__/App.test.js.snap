// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App  tests Snapshot tests 1`] = `
<Component
  basename=""
>
  <nav
    className="navbar navbar-expand-lg navbar-light bg-light"
  >
    <span
      className="navbar-brand"
    >
      AppDev RAS/C Deploy Status
    </span>
    <div
      className="navbar"
      id="navbarNav"
    >
      <ul
        className="navbar-nav"
      >
        <li
          className="nav-item"
        >
          <NavLink
            activeClassName="active"
            className="nav-link"
            exact={true}
            to="/deploys"
          >
            Deploys
          </NavLink>
        </li>
        <li
          className="nav-item"
        >
          <NavLink
            activeClassName="active"
            className="nav-link"
            to="/timeline"
          >
            Timeline
          </NavLink>
        </li>
        <li
          className="nav-item"
        >
          <a
            className="nav-link"
            href="https://docs.google.com/spreadsheets/d/1234567890/"
            rel="noreferrer noopener"
            target="_blank"
          >
            Google Sheet
          </a>
        </li>
      </ul>
    </div>
  </nav>
  <main>
    <Switch>
      <Route
        path="/deploys//startDate/:startDate/endDate/:endDate"
        render={[Function]}
      />
      <Route
        path="/deploys/:appGroup/startDate/:startDate/endDate/:endDate"
        render={[Function]}
      />
      <Route
        path="/deploys/:appGroup?"
        render={[Function]}
      />
      <Route
        path="/timeline//startDate/:startDate/endDate/:endDate"
        render={[Function]}
      />
      <Route
        path="/timeline/:appGroup/startDate/:startDate/endDate/:endDate"
        render={[Function]}
      />
      <Route
        path="/timeline/:appGroup?"
        render={[Function]}
      />
      <Route
        exact={true}
        path="/"
      >
        <Redirect
          to="/deploys"
        />
      </Route>
      <Route
        path="/"
      >
        <FourOhFour />
      </Route>
    </Switch>
  </main>
</Component>
`;
